architecture/additional_concepts/authentication.adoc

```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
 $ oc sa get-token <serviceaccount name>
 ----
 
-* `redirect_uri` must match an annotation on the service account that has the
-prefix `serviceaccounts.openshift.io/oauth-redirecturi`. For example,
-`serviceaccounts.openshift.io/oauth-redirecturi.one`.
 * To get `WWW-Authenticate` challenges, set an
 `serviceaccounts.openshift.io/oauth-want-challenges` annotation on the service
 account to *true*.
 
+* `redirect_uri` must match an annotation on the service account:
+
+
+Annotation keys must have the prefix
+`serviceaccounts.openshift.io/oauth-redirecturi.` or
+`serviceaccounts.openshift.io/oauth-redirectreference.` such as:
+
+----
+serviceaccounts.openshift.io/oauth-redirecturi.<name>
+----
+
+In its simplest form, the annotation can be used to directly specify valid
+redirect URIs. For example:
+
+----
+"serviceaccounts.openshift.io/oauth-redirecturi.first":  "https://example.com"
+"serviceaccounts.openshift.io/oauth-redirecturi.second": "https://other.com"
+----
+
+The `first` and `second` postfixes in the above example are used to separate the
+two valid redirect URIs.
+
+In more complex configurations, static redirect URIs may not be enough. For
+example, perhaps you want all ingresses for a route to be considered valid. This
+is where dynamic redirect URIs via the
+`serviceaccounts.openshift.io/oauth-redirectreference.` prefix come into play.
+
+For example:
+
+----
+"serviceaccounts.openshift.io/oauth-redirectreference.first": "{\"kind\":\"OAuthRedirectReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"Route\",\"name\":\"jenkins\"}}"
+----
+
+Since the value for this annotation contains serialized JSON data, it is easier
+to see in an expanded format:
+
+----
+
+{
+  "kind": "OAuthRedirectReference",
+  "apiVersion": "v1",
+  "reference": {
+    "kind": "Route",
+    "name": "jenkins"
+  }
+}
+
```

Comment:
Actually I misspoke - I used the universal decoder so the server will parse JSON or YAML - it is just really hard to write correct YAML as one giant single line serialized string since whitespace is significant.